Feature/test accessmanager

MikeNeilson opened the pull request:
Initial code for setting up an AccessManager for the test environment.

MikeNeilson committed:
[access] migrated current manager to ServiceProviderInterface.

MikeNeilson committed:
example docker compose file.

MikeNeilson committed:
initial docker-compose file to assist development.

MikeNeilson committed:
initial creation of docker compose file.

MikeNeilson commented:
Converted to draft, the integration tests now fail as I hadn't migrated the auth info in the tests.

MikeNeilson committed:
WIP on initial setup of new login manager.

MikeNeilson committed:
remove attempt at SPI in test code.

MikeNeilson committed:
start of KeyAccessManager.

MikeNeilson committed:
WIP on access managers.

MikeNeilson committed:
General workflow works, database itself needs tweaks.

MikeNeilson committed:
continued work on new access manager and docker-compose

MikeNeilson committed:
User now set correctly from API key.

Created docker file that works from root directory with docker-compose
to simplify development.

TODO:
- retrieve username from connection/datasource.
- figure out why it still says "NO WRITE PRIVS".

MikeNeilson committed:
Additional automated analysis.

MikeNeilson committed:
API Key system works.

NO_WRITE_PRIV was due to getDslContext always adding a set office to the
context based office.

MikeNeilson committed:
API Key considers roles.

MikeNeilson committed:
Works against existing Tomcat CAC auth.

rma-rripken commented on cwms_radar_api/src/main/java/cwms/radar/datasource/DelegatingConnectionPreparer.java:
```diff
@@ -28,6 +38,7 @@ public DelegatingConnectionPreparer(ConnectionPreparer prep1, ConnectionPreparer
     public Connection prepare(Connection connection) throws SQLException {
         Connection retval = connection;
         for (ConnectionPreparer delegate : delegates) {
+            logger.info(delegate.getClass().getName());
```
I think this info() statement should be something quieter.

MikeNeilson replied:
Yep. Missed one. Thanks.

MikeNeilson committed:
correct logging level.

rma-rripken commented on cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java:
```diff
@@ -0,0 +1,195 @@
+
+    private String checkKey(String key, Context ctx) {
+        try(Connection conn = dataSource.getConnection();
+            PreparedStatement setApiUser = conn.prepareStatement("begin cwms_env.set_session_user_direct(upper('q0hectest_pu')); end;");
```
q0hectest_pu seems like it should be pulled out into a constant. Is that a new user for auth checks?

MikeNeilson replied:
oy, thanks, that actually needs to be a "select user from dual" call.

rma-rripken commented on cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java:
```diff
@@ -0,0 +1,195 @@
+            PreparedStatement checkForKey = conn.prepareStatement("select userid from cwms_20.at_api_keys where apikey = ?")) 
+        {
+            setApiUser.execute();
+            checkForKey.setString(1,key);
```
I'll be curious to see all the db interactions that are happening for each request

MikeNeilson replied:
So far it's seemed pretty lightweight, if it starts to add up we can implement a form of session system were anyone can use what ever credentials have creds that don't need to round trip to the database for basic checks.

MikeNeilson committed:
Security Indicator correctly shows up.

Limited to other than GET handlers.
Extended Access Manager to allow implementation specification of the
login Scheme for OpenAPI.

Start on Login.gov/OpenID access manger.

MikeNeilson committed:
Clob write test fails correctly under the new auth system.

MikeNeilson committed:
test now fails for the right reason.

MikeNeilson committed:
Appears to be working correctly.

rma-rripken commented on cwms_radar_api/build.gradle:
```diff
@@ -237,6 +252,7 @@ task integrationTests(type: Test) {
     jvmArgs += "-Djava.util.logging.config.file=$projectDir/logging.properties"
     jvmArgs += "-Dorg.apache.tomcat.util.digester.PROPERTY_SOURCE=org.apache.tomcat.util.digester.EnvironmentPropertySource"
     jvmArgs += "-Dradar.access.provider=KeyAccessManager"
+    jvmArgs += "-Doracle.jdbc.Trace=true"
```
Is this needed or left over?    I haven't played around with this -D, what does it do?  -Doracle.jdbc.Trace=true

MikeNeilson replied:
it was supposed to give more jdbc info to figure something out but it didn't work.

MikeNeilson commented:
That’s left over. It was supposed to give me more logging from the oracle drive as I was trying to track down an error. Need to remove it.


Also, this is still a draft PR, found additional database fixes are required.

Mike

From: rma-rripken ***@***.***>
Sent: Wednesday, January 25, 2023 12:44 PM
To: USACE/cwms-radar-api ***@***.***>
Cc: Mike Neilson ***@***.***>; Author ***@***.***>
Subject: [URL Verdict: Neutral][Non-DoD Source] Re: [USACE/cwms-radar-api] Feature/test accessmanager (PR #238)


@rma-rripken commented on this pull request.

________________________________

In cwms_radar_api/build.gradle<https://github.com/USACE/cwms-radar-api/pull/238#discussion_r1087146915>:

> @@ -237,6 +252,7 @@ task integrationTests(type: Test) {

     jvmArgs += "-Djava.util.logging.config.file=$projectDir/logging.properties"

     jvmArgs += "-Dorg.apache.tomcat.util.digester.PROPERTY_SOURCE=org.apache.tomcat.util.digester.EnvironmentPropertySource"

     jvmArgs += "-Dradar.access.provider=KeyAccessManager"

+    jvmArgs += "-Doracle.jdbc.Trace=true"

Is this needed or left over? I haven't played around with this -D, what does it do? -Doracle.jdbc.Trace=true

—
Reply to this email directly, view it on GitHub<https://github.com/USACE/cwms-radar-api/pull/238#pullrequestreview-1270094510>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AB44KCCK6CJFDUST6EIUOW3WUGGABANCNFSM6AAAAAARCTOAOE>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>

MikeNeilson committed:
Authorization sequence works.

MikeNeilson committed:
Modified how content types are prepared. Part of #272.

MikeNeilson committed:
Auth Test works with bypass database.

Disabled retrieve on create for ClobControllerTestIT. There is an issue
with path parameter handling.

MikeNeilson committed:
baseline access manager tests work.

MikeNeilson committed:
disabled test to fix in later.

MikeNeilson commented:
Okay, got the primary work done, if using the KeyAccessManager the latest unreleased (but is available under the latest-dev tag in the schema installer) is required due to the addition of some new functions and permission tweaks.

Next on the docket is a "combined access manager" and fixing up the rest of the tests. but this is too much change to wait for all that to be done.

rma-rripken commented on cwms_radar_api/src/main/java/cwms/radar/api/ClobController.java:
```diff
@@ -260,14 +262,17 @@ public Clob deserialize(String body, String formatHeader) throws JsonProcessingE
             om = JsonV2.buildObjectMapper();
         } else {
-            throw new FormattingException("Format is not currently supported for Levels");
+            FormattingException fe = new FormattingException("Format is not currently supported for Levels");
```
message should be for Clob not Levels

MikeNeilson replied:
Do'h, I had caught one other of those but missed that one.

MikeNeilson replied:
one of these days we should really move those to a properties file or something, like 90% of that text is duplicate in several controllers.

MikeNeilson committed:
Corrected error message.

rma-rripken commented on cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java:
```diff
@@ -153,14 +147,15 @@ private Set<RouteRole> getRoles(Principal principal) {
                 // It's a CwmsUserPrincipal but it's not our CwmsUserPrincipal.
                 roleNames = callGetRolesReflectively(principal);
-            }
+            }*/
```
Just remove the the try/catch if you're commenting out everything in the catch branch.   Or replace the reflexive stuff with a WARNING message that the cwms_aaa jar needs to be in the tomcat lib dir or whatever the specific fix is.

rma-rripken commented on cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java:
```diff
@@ -169,7 +164,7 @@ private Set<RouteRole> getRoles(Principal principal) {
         return retval;
     }
-
+/* 
```
just delete it if you're /* */ commenting it out.

MikeNeilson replied:
yep, totally forgot about that section.

MikeNeilson committed:
Remove commented out code.

rma-rripken commented on cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java:
```diff
@@ -0,0 +1,203 @@
+    private Set<RouteRole> getRoles(String user,String office) {
+        Set<RouteRole> roles = new HashSet<>();
+        try(Connection conn = dataSource.getConnection();
+            PreparedStatement setApiUser = conn.prepareStatement("begin cwms_env.set_session_user_direct(upper(?)); cwms_env.set_session_office_id(?);end;");
```
you pulled the other statements into constants, why not all of them?

MikeNeilson replied:
would you believe, laziness. But good to call that our. I will adjust.

MikeNeilson replied:
good call, managed to leave a hardcoded office in one instance.

rma-rripken replied:
Submit those fixes and this PR is prob good

MikeNeilson committed:
Added error message to CwmsAccessManasger::getRoles in case classcast exception happened.

rma-rripken commented on cwms_radar_api/src/test/java/fixtures/TomcatServer.java:
```diff
@@ -121,6 +119,11 @@ public Realm getRealm(){
      */
     public void start() throws LifecycleException {
         tomcatInstance.start();
+        try {
```
why is the sleep needed?  Can you add a comment on what happens if the sleep isn't there or how to see if 5 seconds isn't long enough?

MikeNeilson replied:
actually I don't think it is. something was going on that was confusing me but if I remember right it was some other obvious thing like I had a file name slightly wrong so something wasn't loading.

rma-rripken commented on cwms_radar_api/src/test/java/fixtures/TomcatServer.java:
```diff
@@ -77,24 +76,23 @@ public TomcatServer(final String baseDir,
-        Context context = tomcatInstance.addWebapp(contextName, radar.toURI().toURL());
+        //Context context = tomcatInstance.addWebapp(contextName, radar.toURI().toURL());
 
         if(authValve != null && realm != null)
```
Can authValve just go away?  Seems like we don't use it anymore?

MikeNeilson replied:
yeah, but before I go that far I want to clean up other things.

rma-rripken commented on cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java:
```diff
@@ -0,0 +1,203 @@
+            logger.log(Level.WARNING,"Unauthorized login attempt",ex);
+            HashMap<String,String> msg = new HashMap<>();
+            msg.put("message",ex.getMessage());
+            RadarError re = new RadarError("Unauthorized",msg,true);
```
There are minor formatting/spacing issues in this file

rma-rripken commented on cwms_radar_api/src/main/java/cwms/radar/formatters/ContentType.java:
```diff
@@ -46,17 +47,30 @@ public String getCharset() {
+     * For us everything else is informational or used indirectly
+     */
     @Override
     public boolean equals(Object other) {
         logger.atFinest().log("Checking %s vs %s", this, other);
         if (!(other instanceof ContentType)) return false;
         ContentType o = (ContentType) other;
         if (!(mediaType.equals(o.mediaType))) return false;
 
+        /** We loop through instead of using contains key. 
+         *  Content-type parameter names are not case sensitive.
+         */
         for (Map.Entry<String, String> entry : parameters.entrySet()) {
             String key = entry.getKey();
-            if (key.equals("q")) continue; // we don't care about q for equals
-            if (!entry.getValue().equals(o.parameters.get(key))) return false;
+            if( "version".equalsIgnoreCase(key) ) {
+                return entry.getValue().equals(o.parameters.get(key));
+            }
```
I think this change deserves some tests in ContentTypeTest to demonstrate what we expect to be equals and what isn't.

MikeNeilson replied:
simple test added. I'll grant the names may not be the best but for the moment I was going for change the least amount of code at one time.

rma-rripken commented on cwms_radar_api/src/main/java/cwms/radar/formatters/ContentType.java:
```diff
@@ -88,4 +102,18 @@ public String toString() {
+     * @return whether they are equivalent
+     */
+    final public static boolean equivalent(String a, String b) {
+        Objects.requireNonNull(a, "Cannot determine equivalency of null content-types");
```
Needs unit test.  tests should also demonstrate how this differs from equals

rma-rripken commented on cwms_radar_api/src/test/java/fixtures/RadarApiSetupCallback.java:
```diff
@@ -35,11 +39,11 @@ public class RadarApiSetupCallback implements BeforeAllCallback,AfterAllCallback
+    private static final String ORACLE_IMAGE = System.getProperty("RADAR.oracle.database.image", CwmsDatabaseContainer.ORACLE_19C);
     private static final String ORACLE_VOLUME = System.getProperty("RADAR.oracle.database.volume", "cwmsdb_radar_volume");
-    private static final String CWMS_DB_IMAGE = System.getProperty("RADAR.cwms.database.image", "registry.hecdev.net/cwms_schema_installer:21.1.1");
+    private static final String CWMS_DB_IMAGE = System.getProperty("RADAR.cwms.database.image", "registry.hecdev.net/cwms/schema_installer:latest-dev");
+
```
Does this mean we're always running against the latest version of the installer or the latest version of the schema or both?  I think I'd rather depend on a specific version than latest.

MikeNeilson replied:
That does mean the latest schema installer (basically what's in master). The tests currently do depend on it. But we should change it once a release that includes the changes is made.

MikeNeilson committed:
fix general formatting issues.

MikeNeilson committed:
remove unrequired sleep.

MikeNeilson committed:
updated comments and added tests for ContentType.

MikeNeilson committed:
corrected query to have variable.
